pkg/compactor/blocks_cleaner_test.go

```diff
@@ -681,6 +686,95 @@ func TestBlocksCleaner_ShouldRemoveBlocksOutsideRetentionPeriod(t *testing.T) {
 	}
 }
 
+func TestBlocksCleaner_ShouldRemovePartialBlocksOutsideDelayPeriod(t *testing.T) {
+	bucketClient, tempDir := mimir_testutil.PrepareFilesystemBucket(t)
+	bucketClient = bucketindex.BucketWithGlobalMarkers(bucketClient)
+
+	ts := func(hours int) int64 {
+		return time.Now().Add(time.Duration(hours)*time.Hour).Unix() * 1000
+	}
+
+	block1 := createTSDBBlock(t, bucketClient, "user-1", ts(-10), ts(-8), 2, nil)
+	block2 := createTSDBBlock(t, bucketClient, "user-1", ts(-8), ts(-6), 2, nil)
+
+	cfg := BlocksCleanerConfig{
+		DeletionDelay:           time.Hour,
+		CleanupInterval:         time.Minute,
+		CleanupConcurrency:      1,
+		DeleteBlocksConcurrency: 1,
+	}
+
+	ctx := context.Background()
+	logger := test.NewTestingLogger(t)
+	reg := prometheus.NewPedanticRegistry()
+	cfgProvider := newMockConfigProvider()
+
+	cleaner := NewBlocksCleaner(cfg, bucketClient, tsdb.AllUsers, cfgProvider, logger, reg)
+
+	assertBlockExists := func(user string, block ulid.ULID, expectExists bool) {
+		exists, err := bucketClient.Exists(ctx, path.Join(user, block.String(), metadata.MetaFilename))
+		require.NoError(t, err)
+		assert.Equal(t, expectExists, exists)
+	}
+
+	makeBlockPartial := func(user string, block ulid.ULID) {
+		err := os.Remove(path.Join(tempDir, user, block.String(), metadata.MetaFilename))
+		require.NoError(t, err)
+	}
+
+	assertBlockExists("user-1", block1, true)
+	assertBlockExists("user-1", block2, true)
+	makeBlockPartial("user-1", block1)
+	assertBlockExists("user-1", block1, false)
+	assertBlockExists("user-1", block2, true)
+	require.NoError(t, cleaner.cleanUser(ctx, "user-1"))
+
+	// Set partial block delay such that block will not be marked for deletion
+	// The comparison is based on inode modification time, so anything more than very recent (< 1 second) won't be
+	// out of range
+	cfgProvider.userPartialBlockDelay["user-1"] = 1 * time.Hour
+	require.NoError(t, cleaner.cleanUser(ctx, "user-1"))
+
+	assert.NoError(t, testutil.GatherAndCompare(reg, strings.NewReader(`
```

Comment:
The test shouldn't continue after an error here:
```suggestion
	require.NoError(t, testutil.GatherAndCompare(reg, strings.NewReader(`
```